### DISCUSSION TITLE
Cannot build main with maven

### DISCUSSION THREAD

**Discussion by fkleedorfer** [eclipse-rdf4j/rdf4j#3230] [Q&A] [answered]
First time I am trying to build rdf4j. I just cloned the repo and went on to build using 

```
mvn install
```
One of the tests fails:
```
[INFO] Running org.eclipse.rdf4j.rio.ndjsonld.NDJSONLDWriterTest
[ERROR] Tests run: 60, Failures: 0, Errors: 1, Skipped: 2, Time elapsed: 1.198 s <<< FAILURE! - in org.eclipse.rdf4j.rio.ndjsonld.NDJSONLDWriterTest
[ERROR] org.eclipse.rdf4j.rio.ndjsonld.NDJSONLDWriterTest.testRoundTrip_NonDefaultCharEncoding  Time elapsed: 0.004 s  <<< ERROR!
org.eclipse.rdf4j.rio.RDFParseException: Could not parse JSONLD [line 1, column 83]
Caused by: com.fasterxml.jackson.databind.JsonMappingException:
Invalid UTF-8 start byte 0xfc
 at [Source: (ByteArrayInputStream); line: 1, column: 83] (through reference chain: java.util.ArrayList[0])
Caused by: com.fasterxml.jackson.core.JsonParseException:
Invalid UTF-8 start byte 0xfc
 at [Source: (ByteArrayInputStream); line: 1, column: 83]

[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   NDJSONLDWriterTest>RDFWriterTest.testRoundTrip_NonDefaultCharEncoding:353 ▒ RDFParse
[INFO]
[ERROR] Tests run: 327, Failures: 0, Errors: 1, Skipped: 7
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  9.773 s
[INFO] Finished at: 2021-07-25T14:09:16+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M5:test (default-test) on project rdf4j-rio-jsonld: There are test failures.
```

maven versions 3.6.3 and 3.8.1 (tried both)
java: AdoptOpenJDK 15
OS: windows 10
running in git bash

Interestingly, the test passes when I run it via Intellij. 

What am I missing?

**Comment by hmottestad** [comment]
Nifty. Looks like a file encoding issue. I haven't looked into that test, but from the test name I guess it tests the system local file encoding.

**Reply by abrokenjester** [reply]
The odd thing is that the test itself _explicitly_ uses a non-standard encoding (windows-1250) to write, then read a string. That should not be be platform-dependent, but clearly it is.

**Comment by barthanssens** [comment]
Interesting, trying to reproduce this on Win10 R2, with AdoptOpenJDK (16 instead of 15) OpenJ9 0.26, maven 3.6.3 in git 2.31.1 command shell.
I get another issue

```
import: Entry[import  from realm ClassRealm[project>org.eclipse.rdf4j:rdf4j:3.7.2-SNAPSHOT, parent: ClassRealm[maven.api, parent: null]]]
[ERROR]
[ERROR] -----------------------------------------------------
[ERROR] : Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @a3340328
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginContainerException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :rdf4j-model-api
```

**Reply by abrokenjester** [reply]
I'm not sure how you can run _both_ AdoptOpenJDK 16 and OpenJ9, but fwiw on my (linux) machine using OpenJDK 16 and maven 3.8.1, everything builds as expected. 

What is your `$JAVA_HOME` env variable set to? That's what maven uses to determine which JDK to use. And I've noticed that maven 3.6.3 seems to struggle in combination with newer Java versions.

**Reply by barthanssens** [reply]
I'll look into it, I'm almost sure it is yet again a shady setting on my machine ;-)

FWIW,  AdoptOpenJDK comes in two flavors, using the same OpenJDK class libraries but a different VM (HotSpot or OpenJ9)

**Reply by barthanssens** [reply]
See #3253

**Comment by abrokenjester** [comment, marked as answer]
Thanks for reporting this. I think this is a bug in the `NDJSONLDParser` that only manifests itself when running the test suite on a Windows machine. The root cause of the issue is, I suspect [line 63](https://github.com/eclipse/rdf4j/blob/65071ebf42c5accc09081f4959a455926f0c4cc2/core/rio/jsonld/src/main/java/org/eclipse/rdf4j/rio/ndjsonld/NDJSONLDParser.java#L63):

```java
JsonParser nextParser = factory.createParser(new ByteArrayInputStream(line.getBytes()));
```

In particular, the `line.getBytes()` has no encoding supplied, so it falls back to the default platform encoding (which on Windows is not UTF-8 but something like windows-1250, which the underlying JSON parser doesn't expect). 

To fix, upgrade your system to Linux.

I kid. I'll log a bug ticket for this. I don't think the fix is hard but it needs a careful eye (I'm also a bit suspicious of the fact that that protected method override for `getJSONObject` completely ignores the `InputStream in` argument and just only uses the supplied reader).

**Reply by abrokenjester** [reply]
Issue logged as #3252

**Reply by JervenBolleman** [reply]
This should be fixed by [JEP-400](https://openjdk.java.net/jeps/400) but that is java-18+ so it is a good thing #3252 is fixed :)
